bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/handler/OpenWebNetBridgeHandler.java

```diff
@@ -0,0 +1,566 @@
+            if (discoveryConfig != null) {
+                if (discoveryConfig instanceof java.lang.Boolean) {
+                    discoveryByActivation = (boolean) getConfig().get(CONFIG_PROPERTY_DISCOVERY_ACTIVATION);
+                    logger.debug("discoveryByActivation={}", discoveryByActivation);
+                } else {
+                    logger.warn(
+                            "invalid discoveryByActivation parameter value (should be true/false). Keeping current value={}.",
+                            discoveryByActivation);
+                }
+            }
+            logger.debug("Creating new BUS gateway with config properties: {}:{}, pwd={}", host, port, passwdMasked);
+            return new BUSGateway(host, port, passwd);
+        } else {
+            logger.warn("Cannot connect to gateway. No host/IP has been provided in Bridge configuration.");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
+                    "@text/offline.conf-error-no-ip-address");
+            return null;
+        }
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.debug("handleCommand (command={} - channel={})", command, channelUID);
+        OpenGateway gw = gateway;
+        if (gw != null && !gw.isConnected()) {
+            logger.warn("BridgeHandler gateway is NOT connected, skipping command");
+            return;
+        } else {
+            logger.warn("Channel not supported: channel={}", channelUID);
+        }
+    }
+
+    @Override
+    public Collection<ConfigStatusMessage> getConfigStatus() {
+        Collection<ConfigStatusMessage> configStatusMessages;
+        configStatusMessages = Collections.emptyList();
+        return configStatusMessages;
+    }
+
+    @Override
+    public void thingUpdated(Thing thing) {
+        super.thingUpdated(thing);
+        logger.info("Bridge configuration updated.");
+        // for (Thing t : getThing().getThings()) {
+        // final ThingHandler thingHandler = t.getHandler();
+        // if (thingHandler != null) {
+        // thingHandler.thingUpdated(t);
+        // }
+        // }
+    }
+
+    @Override
+    public void handleRemoval() {
+        disconnectGateway();
+        super.handleRemoval();
+    }
+
+    @Override
+    public void dispose() {
+        disconnectGateway();
+        super.dispose();
+    }
+
+    private void disconnectGateway() {
+        OpenGateway gw = gateway;
+        if (gw != null) {
+            gw.closeConnection();
+            gw.unsubscribe(this);
+            logger.debug("gateway {} connection closed and unsubscribed", gw.toString());
+            gateway = null;
+        }
+        reconnecting = false;
+    }
+
+    /**
+     * Search for devices connected to this bridge handler's gateway
+     *
+     * @param listener to receive device found notifications
+     */
+    public synchronized void searchDevices() {
```

Comment:
I would have expected most of this discovery code to be in the discovery service. To keep code better separated. Or is there a reason the code is here in the handler?

Reply:
`searchDevices `and `onNewDevice`/`onDiscoveryCompleted `are here because it's `OpenWebNetBridgeHandler ` that owns the communication to the BTicino gateway (by using openwebnet protocol), through which device discovery is performed.